## https://adventofcode.com/2024/day/3

In [5]:
import pandas as pd
import numpy as np
import re

pattern = r"mul\((\d+),(\d+)\)"

matches = []

with open('../Data_Inputs/AoC_3_input.txt', 'r') as file:
    for line in file:
        # matches.extend(re.findall(pattern, line))
        matches.extend(re.findall(pattern, line))

# for i, match in enumerate(matches):
#     print(f"arr[{i}] = {match}")

In [6]:
print(matches[0])
type(matches[0])

('402', '190')


tuple

In [15]:
print(matches[0][1])
type(matches[0][0])

190


str

In [16]:
sum = 0
for x, y in matches:
    sum += (int(x) * int(y))

In [17]:
print(sum)

183788984


In [ ]:
import re

pattern = re.compile("do\(\)")

enabled = []
with open('../Data_Inputs/AoC_3_input.txt', 'r') as file:
    for line in file:
        # print(line)
        m = pattern.search(line)
        print(m.start())
        print(m.end())
        # print(m)


126
130


In [5]:
import re

pattern = re.compile("do\(\)")

enabled = []
with open('../Data_Inputs/AoC_3_input.txt', 'r') as file:
    for line in file:
        # print(line)
        m = pattern.search(line)
        print(m.start())
        print(m.end())
        # print(m)

458
462
103
107
798
802
299
303
570
574
126
130


In [ ]:
with open('../Data_Inputs/AoC_3_input.txt', 'r') as file:
    full_string = ''.join(line.strip() for line in file)

print(full_string)

In [8]:
import re

pattern = re.compile("(?<=do\(\))(.*?)(?=don't\(\))")


with open('../Data_Inputs/AoC_3_input.txt', 'r') as file:
    for line in file:
        m = pattern.search(line)
        if m:
            print(m.group(1))

@*{mul(557,916)&(who()mul(445,933):(,;>?mul(108,268)
+how()![where()mul(399,972)mul(862,828)when()?^&)mul(200,579)how()who()mul(79,535) <
<}$%;<>how()'mul(460,761)(mul(83,660)[where()mul(539,808'&select()*
how()>mul(24,375)'!;,,mul(34,942)?-mul(511,814)<:{{how()}{what()mul(369,392)mul(643,828)mul(12,913)from()~select(){ +$$mul(384,257){who()!{$ )#/!mul(607,455)how()*(( /-+::mul(747,552)mul(333,364)how()*?select(914,438)% ~,mul(257,957)do()where(){+#mul(85,930)when()^'<~^/ mul(538,737)?,%what()*-^ where()mul(959,252)mul(435,422)#where()mulwhere()mul(634,612),}{select(863,784){[do()who()}*+mul(842,759),#what());[}why(616,35)%mul(631,814);where()/)mul(768,481)>{!)>who()who()select()how()mul(432,507)&)!>{%>(mul:$?(?~@ }mul(60,211)what()what()*what()when()+];mul(960,9)- ~'mul(678]+&!mul(252,399)what()?:?} !@>[mul(510,368)mul(674,905)how()/why()^mul(763,579)mul(771,675)mul(565,944)'who(345,580)#<where(248,141)from()
#?what()where()mul(351,931)>]%why()&mul(809,5~*/&&why()>mul(571,244)where()#

In [ ]:
import re

pattern = re.compile("(?<=do\(\))(.*?)(?=don't\(\))")
mul_pattern = re.compile("mul\((\d+),(\d+)\)")

matches = []

with open('../Data_Inputs/AoC_3_input.txt', 'r') as file:
    for line in file:
        m = pattern.search(line)
        if m:
            matches.extend(mul_pattern.findall(m.group(1)))

In [16]:
print(matches[3])
print(matches[1][1])

('399', '972')
933


In [ ]:
sum = 0
for x, y in matches:
    sum += (int(x) * int(y))

print(sum)
# answer given is 13904462. This is incorrect and is lower than the actual answer.

13904462


In [17]:
for match in matches:
    print(match)

('557', '916')
('445', '933')
('108', '268')
('399', '972')
('862', '828')
('200', '579')
('79', '535')
('460', '761')
('83', '660')
('24', '375')
('34', '942')
('511', '814')
('369', '392')
('643', '828')
('12', '913')
('384', '257')
('607', '455')
('747', '552')
('333', '364')
('257', '957')
('85', '930')
('538', '737')
('959', '252')
('435', '422')
('634', '612')
('842', '759')
('631', '814')
('768', '481')
('432', '507')
('60', '211')
('960', '9')
('252', '399')
('510', '368')
('674', '905')
('763', '579')
('771', '675')
('565', '944')
('351', '931')
('571', '244')
('518', '122')
('579', '300')
('221', '419')
('693', '760')
('516', '571')
('325', '165')
('346', '3')
('248', '163')
('57', '514')
('871', '508')
('498', '284')
('840', '612')
('897', '30')
('405', '945')
('346', '138')
('354', '691')


## Ah, I forgot to grab the inital mul(x, y) values before the first do() or don't(). Also, this is only grabbing the first pattern match per line in the text. 

In [19]:
import re

pattern = re.compile("(?<=do\(\))(.*?)(?=don't\(\))")
mul_pattern = re.compile("mul\((\d+),(\d+)\)")

matches = []

with open('../Data_Inputs/AoC_3_input.txt', 'r') as file:
    for line in file:
        m = pattern.search(line)
        if m:
            matches.extend(mul_pattern.findall(m.group(1)))

In [27]:
import re

init_pattern = re.compile(".*?(do\(\).*?don't\(\)|don't\(\))")
pattern = re.compile("(?<=do\(\))(.*?)(?=don't\(\))")
mul_pattern = re.compile("mul\((\d+),(\d+)\)")

matches = []
mul_matches = []

with open('../Data_Inputs/AoC_3_input.txt', 'r') as file:
    for line in file:
        matches.extend(pattern.findall(line))
        for match in matches:
            mul_matches.extend(mul_pattern.findall(match))

In [21]:
for match in mul_matches:
    print(match)

('557', '916')
('445', '933')
('108', '268')
('891', '69')
('546', '834')
('767', '879')
('876', '495')
('911', '71')
('116', '462')
('258', '286')
('97', '540')
('682', '75')
('63', '26')
('336', '475')
('637', '386')
('846', '915')
('795', '805')
('370', '651')
('705', '839')
('112', '21')
('285', '252')
('451', '396')
('57', '589')
('554', '575')
('856', '360')
('530', '200')
('588', '2')
('604', '72')
('250', '680')
('555', '956')
('377', '822')
('457', '538')
('52', '298')
('452', '941')
('557', '916')
('445', '933')
('108', '268')
('891', '69')
('546', '834')
('767', '879')
('876', '495')
('911', '71')
('116', '462')
('258', '286')
('97', '540')
('682', '75')
('63', '26')
('336', '475')
('637', '386')
('846', '915')
('795', '805')
('370', '651')
('705', '839')
('112', '21')
('285', '252')
('451', '396')
('57', '589')
('554', '575')
('856', '360')
('530', '200')
('588', '2')
('604', '72')
('250', '680')
('555', '956')
('377', '822')
('457', '538')
('52', '298')
('452', '941')
('39

## Okay, I think I have it fixed to find all of the pattern matches, but now to also grab the inital mul(x, y) matches

In [ ]:
import re

init_pattern = re.compile(".*?(do\(\).*?don't\(\)|don't\(\))")
pattern = re.compile("(?<=do\(\))(.*?)(?=don't\(\))")
mul_pattern = re.compile("mul\((\d+),(\d+)\)")

matches = []
mul_matches = []

full_string = ""

with open('../Data_Inputs/AoC_3_input.txt', 'r') as file:
    full_string = ''.join(line.strip() for line in file)


print(full_string)

In [37]:
test = (init_pattern.search(full_string))
print(test.group(0))

$  mul(402,190))&<why(211,617)how()/;mul(506,313)[^^<!$#when(636,198),]mul(744,268)#&!what()&!;mul(631,641)where()@select()!@?(:;how()mul(25,288)&~^select()mul(683,657)'mul(505,315)+~:-mul(671,741)]{mul(397,54))?mul(511,935)where()select()where()-mul(525,623)mul(206,770){/}don't()


In [38]:
test_list = []
test_list.extend(mul_pattern.findall(test.group(0)))

In [39]:
print(test_list[0])

('402', '190')


In [ ]:
import re

init_pattern = re.compile(".*?(do\(\).*?don't\(\)|don't\(\))")
pattern = re.compile("(?<=do\(\))(.*?)(?=don't\(\))")
mul_pattern = re.compile("mul\((\d+),(\d+)\)")

matches = []
mul_matches = []
first_line = True

with open('../Data_Inputs/AoC_3_input.txt', 'r') as file:
    for line in file:
        if first_line:
            init = init_pattern.search(line)
            mul_matches.extend(mul_pattern.findall(init.group(0)))
        matches.extend(pattern.findall(line))
        for match in matches:
            mul_matches.extend(mul_pattern.findall(match))

In [42]:
print(mul_matches[0])

('402', '190')


In [ ]:
sum = 0
for x, y in mul_matches:
    sum += (int(x) * int(y))

print(sum)
# answer of 165703245 is incorrect and too high

165703245


In [4]:
for match in mul_matches:
    print(match)

('402', '190')
('506', '313')
('744', '268')
('631', '641')
('25', '288')
('683', '657')
('505', '315')
('671', '741')
('397', '54')
('511', '935')
('525', '623')
('206', '770')
('557', '916')
('445', '933')
('108', '268')
('891', '69')
('546', '834')
('767', '879')
('876', '495')
('911', '71')
('116', '462')
('258', '286')
('97', '540')
('682', '75')
('63', '26')
('336', '475')
('637', '386')
('846', '915')
('795', '805')
('370', '651')
('705', '839')
('112', '21')
('285', '252')
('451', '396')
('57', '589')
('554', '575')
('856', '360')
('530', '200')
('588', '2')
('604', '72')
('250', '680')
('555', '956')
('377', '822')
('457', '538')
('52', '298')
('452', '941')
('557', '916')
('445', '933')
('108', '268')
('891', '69')
('546', '834')
('767', '879')
('876', '495')
('911', '71')
('116', '462')
('258', '286')
('97', '540')
('682', '75')
('63', '26')
('336', '475')
('637', '386')
('846', '915')
('795', '805')
('370', '651')
('705', '839')
('112', '21')
('285', '252')
('451', '396')
(

## i'm so dumb, I forgot to change first_line to false after it processes the first time. That didn't fix the issue. For some reason, I was repeating lines in the loop. I will now convert the parsed file to a single string with no line breaks.

In [11]:
import re

init_pattern = re.compile(".*?(do\(\).*?don't\(\)|don't\(\))")
pattern = re.compile("(?<=do\(\))(.*?)(?=don't\(\))")
mul_pattern = re.compile("mul\((\d+),(\d+)\)")

matches = []
mul_matches = []
first_line = True
full_string = ""

with open('../Data_Inputs/AoC_3_input.txt', 'r') as file:
    full_string = ''.join(line.strip() for line in file)

if first_line:
    init = init_pattern.search(full_string)
    mul_matches.extend(mul_pattern.findall(init.group(0)))
    # for match in mul_matches:
    #     print(match)
    first_line = False
matches.extend(pattern.findall(full_string))
# for match in matches:
#     print(match)

for match in matches:
    mul_matches.extend(mul_pattern.findall(match))
#     for match in mul_matches:
#         print(match)

In [12]:
sum = 0
for x, y in mul_matches:
    sum += (int(x) * int(y))

print(sum)
# answer of 140933863 is incorrect and too high

62098619


## Yes, this was the issue and 62098619 is the correct answer